### PR TITLE
feat: double_quotes

### PR DESCRIPTION
**Pull request by simojoe**
This pull request adds a new rule to `wdl`.

- **Rule Name**: `double_quotes`
- **Rule Kind**: Lint warning
- **Rule Code**: `v1::W012`
- **Packages**: `wdl-grammar`

Warns about strings defined with single-quotes. This ignores anything located inside the command section. Tests are not exhaustive, there are quite a lot of things that are using strings. Should they still be added?

Before submitting this PR, please make sure:

- [X] You have added a few sentences describing the PR here.
- [ ] You have added yourself or the appropriate individual as the assignee.
- [ ] You have added at least one relevant code reviewer to the PR.
- [X] Your code builds clean without any errors or warnings.
- [X] You have added an entry to the relevant `CHANGELOG.md` (see
      ["keep a changelog"] for more information).
- [X] Your commit messages follow the [conventional commit] style.
- [ ] Your changes are squashed into a single commit (unless there is a _really_
      good, articulated reason as to why there should be more than one).

Rule specific checks:

- [X] You have added the rule as an entry within the the package-specific rule
      tables (`wdl-ast/src/v1.rs` for AST-based rules and 
      `wdl-grammar/src/v1.rs` for parse tree-based rules).
- [X] You have added the rule as an entry within the the global rule
      table at `RULES.md`.
- [X] You have added the rule to the appropriate `fn rules()`.
    - Validation rules added to `wdl-ast` should be added to `fn rules()` within 
      `wdl-ast/src/v1/validation.rs`.
    - Lint rules added to `wdl-ast` should be added to `fn rules()` within `wdl-ast/src/v1/lint.rs`.
    - Validation rules added to `wdl-grammar` should be added to `fn rules()` within 
      `wdl-grammar/src/v1/validation.rs`.
    - Lint rules added to `wdl-grammar` should be added to `fn rules()` within 
      `wdl-grammar/src/v1/lint.rs`.
- [ ] You have added a test that covers every possible setting for the rule 
      within the file where the rule is implemented.
- [X] You have run `wdl-gauntlet --save-config` to ensure that all of the rules
      added/removed are now reflected in the baseline configuration file
      (`Gauntlet.toml`).

[conventional commit]: https://www.conventionalcommits.org/en/v1.0.0/#summary
["keep a changelog"]: https://keepachangelog.com/en/1.0.0/

### CODE REVIEW

File: wdl-grammar/src/v1/lint/double_quotes.rs
Comment (a-frantz) [inline wdl-grammar/src/v1/lint/double_quotes.rs:48]: I'm not sure this falls ender `Naming`? It's definitely `Style` (which is being auto-added bc of the `Naming` tag), and I think this rule calls for a new tag called `Consistency`. And if we're adding a `Consistency` tag, that should be added to a few of the existing rules: `whitespace`, `no_curly_commands`, `mixed_indentation`, `snake_case`? Open to hearing dissenting opinions on any of these specifics here.

**Review by a-frantz — COMMENTED** [review]
Does this work for nested placeholder strings? I haven't checked out the branch to test myself. e.g. `String nested = "Hello ~{if alien then 'world' else planet}!"`. The inner `'world'` string should be flagged. There are other cases we probably want to check as well.

I'm fine with skipping over the `command` block for now. That seems like a bonus check to me, that shouldn't hold up the other test conditions.

File: wdl-grammar/src/v1/lint/double_quotes.rs
Comment (adthrasher) [reply wdl-grammar/src/v1/lint/double_quotes.rs:48]: It is labeled `Naming` in the discussion (https://github.com/stjude-rust-labs/wdl/discussions/11), but I agree. I'm not sure that `Naming` is correct.

File: wdl-grammar/src/v1/lint/double_quotes.rs
Comment (adthrasher) [inline wdl-grammar/src/v1/lint/double_quotes.rs:17]: Does this need a comment?

File: wdl-grammar/src/v1/lint/double_quotes.rs
Comment (adthrasher) [reply wdl-grammar/src/v1/lint/double_quotes.rs:48]: `mixed_indentation` uses `Style|Spacing|Clarity`. Is `Clarity` sufficient here or does `Consistency` convert something different?

Comment (adthrasher) [reply wdl-grammar/src/v1/lint/double_quotes.rs:48]: `snake_case` similarly uses `Naming|Style|Clarity`

Comment (a-frantz) [reply wdl-grammar/src/v1/lint/double_quotes.rs:48]: Hm. They are similar, but different.

Generally, I'd argue "too many tags" is a better problem to have than "not enough tags". Tags serve the (primary?) function of making the docs easier to navigate and the (secondary?) function of providing easy shorthand for groups of rules that a user would want to toggle at once. (For the sake of argument, we will pretend those functions exist today lol)

If those are our two main goals for tags, I don't think there's a real problem with having too many. The biggest downside (I can see) is that having lots of tags muddies the warning's shorthand notation (`[v1::W004::[Spacing, Style, Clarity]::High]`).

Comment (simojoe) [reply wdl-grammar/src/v1/lint/double_quotes.rs:48]: What is the final decision here? Only `Clarity` or `Clarity|Style`?

File: wdl-grammar/src/v1/lint/double_quotes.rs
Comment (simojoe) [reply wdl-grammar/src/v1/lint/double_quotes.rs:17]: Yes, forgot this one. Strangely none of the steps described in the README picked it up but the lint here did.

Will add.

File: wdl-grammar/src/v1/lint/double_quotes.rs
Comment (a-frantz) [reply wdl-grammar/src/v1/lint/double_quotes.rs:48]: I misread @adthrasher 's comments. Let me try addressing this again lol (so @simojoe hold on before making any changes)

>  I would tend to favor `Clarity` here, ironically, for consistency with the already implemented rules.

I don't see how `Clarity` applies to `double_quotes`. This rule isn't a question of "what's clearer?".

As authors, everyone is faced with what is a more or less identical choice between single and double quotes. We've arbitrarily chosen double quotes to be the standard because switching back and forth is bad. But as long as it's consistent, it doesn't matter.

That's why I want to add `Consistency` here. And then in hindsight, `Consistency` _could_ apply to some of our other rules.

What's the argument against having both `Consistency` and `Clarity`?

Comment (adthrasher) [reply wdl-grammar/src/v1/lint/double_quotes.rs:48]: @a-frantz I think it's worth trying to avoid too many tags if we can. When I look at the rules `snake_case`, `no_curly_commands`, and `mixed_indentation`, in all three cases we are choosing a preferred style for the purpose of `Clarity`. Technically, all would be `Consistency` as well.

Comment (a-frantz) [reply wdl-grammar/src/v1/lint/double_quotes.rs:48]: options:
1. no new `Consistency` tag
    * IMO `Clarity` doesn't quite fit for `double_quotes`. Which means `double_quotes` would only have the tag `Style`.
2. replace all instances of `Clarity` with `Consistency`
    * `double_quotes` gets both `Consistency` and `Style`
3. Have both `Clarity` and `Consistency`.
    * They appear together for `snake_case`, `no_curly_commands`, and `mixed_indentation` _but not_ `double_quotes` (which only gets `Consistency`)
 4. ??? is there anything else worth considering?
 
I would rule out number 2. Sounds like you want to rule out number 3. I don't feel strongly about number 1, so maybe that's the compromise we settle on?

Comment (adthrasher) [reply wdl-grammar/src/v1/lint/double_quotes.rs:48]: Copying/summarizing conversation with @a-frantz from Slack, but the conclusion is that `double_quotes` should get `Style` and `Clarity` tags.

Personally, I view `consistency` as a necessary component of `clarity`. If something is inconsistent (e.g. mixing single and double quotes), it lacks clarity. The choice of double quotes vs single quotes comes down to a matter of `Style` in my opinion as well. So `Style|Clarity` seems to fit.

There will also likely be an effort in the future to define a set of tags to support and write descriptions for each. With that in mind, keeping a smaller set of tags now makes that easier. We can always add tags to existing rules as part of that effort.

**Comment by adthrasher** [comment]
> Does this work for nested placeholder strings? I haven't checked out the branch to test myself. e.g. `String nested = "Hello ~{if alien then 'world' else planet}!"`. The inner `'world'` string should be flagged. There are other cases we probably want to check as well.
> 
> I'm fine with skipping over the `command` block for now. That seems like a bonus check to me, that shouldn't hold up the other test conditions.

It does not appear to catch that condition.

**Comment by adthrasher** [comment]
I rebased this to get the latest changes. Looking at the `Arena` results, it does appear to be checking within the `command` block.

**Comment by a-frantz** [comment]
> I rebased this to get the latest changes. Looking at the `Arena` results, it does appear to be checking within the `command` block.

Ok so that's a good thing I think. As long as it doesn't apply to literal Bash code. e.g. look at this hypothetical command block
```
command <<<
    echo 'this Bash string should be ignored'
    echo "~{if foo then 'this should be flagged' else 'this one too'}"
>>>
```

**Comment by adthrasher** [comment]
> > I rebased this to get the latest changes. Looking at the `Arena` results, it does appear to be checking within the `command` block.
> 
> Ok so that's a good thing I think. As long as it doesn't apply to literal Bash code. e.g. look at this hypothetical command block
> 
> ```
> command <<<
>     echo 'this Bash string should be ignored'
>     echo "~{if foo then 'this should be flagged' else 'this one too'}"
> >>>
> ```

To clarify, it filters to `v1::Rule::string` instances. So it will ignore anything that isn't part of a WDL instruction in the `command` block. So in your example, it ignores the first line, but will catch the two embedded strings.

**Comment by a-frantz** [comment]
> > > I rebased this to get the latest changes. Looking at the `Arena` results, it does appear to be checking within the `command` block.
> > 
> > 
> > Ok so that's a good thing I think. As long as it doesn't apply to literal Bash code. e.g. look at this hypothetical command block
> > ```
> > command <<<
> >     echo 'this Bash string should be ignored'
> >     echo "~{if foo then 'this should be flagged' else 'this one too'}"
> > >>>
> > ```
> 
> To clarify, it filters to `v1::Rule::string` instances. So it will ignore anything that isn't part of a WDL instruction in the `command` block. So in your example, it ignores the first line, but will catch the two embedded strings.

That's perfect then. Let's add a my example as an actual test case for sanity, update Arena, and then this should be good to merge.

**Comment by adthrasher** [comment]
> > > > I rebased this to get the latest changes. Looking at the `Arena` results, it does appear to be checking within the `command` block.
> > > 
> > > 
> > > Ok so that's a good thing I think. As long as it doesn't apply to literal Bash code. e.g. look at this hypothetical command block
> > > ```
> > > command <<<
> > >     echo 'this Bash string should be ignored'
> > >     echo "~{if foo then 'this should be flagged' else 'this one too'}"
> > > >>>
> > > ```
> > 
> > 
> > To clarify, it filters to `v1::Rule::string` instances. So it will ignore anything that isn't part of a WDL instruction in the `command` block. So in your example, it ignores the first line, but will catch the two embedded strings.
> 
> That's perfect then. Let's add a my example as an actual test case for sanity, update Arena, and then this should be good to merge.

Test added. I've updated Arena, but I was going to hold off until https://github.com/stjudecloud/workflows/pull/157 is merged. However, I can commit the Arena warnings and move ahead with this first if you prefer.

**Review by a-frantz — APPROVED** [review]

**Pull request merged**
